src/ui/components/SecondaryToolbox/ReduxDevTools.tsx

```diff
@@ -133,6 +138,96 @@ function isFrameInDecl(decl: ApplyMiddlewareDecl, frame: PauseFrame) {
   );
 }
 
+function isNestedInside(child: SourceLocationRange, parent: SourceLocationRange) {
+  const startsBefore =
+    parent.begin.line < child.begin.line ||
+    (parent.begin.line === child.begin.line && parent.begin.column <= child.begin.column);
+  const endsAfter =
+    parent.end.line > child.end.line ||
+    (parent.end.line === child.end.line && parent.end.column >= child.end.column);
+
+  return startsBefore && endsAfter;
+}
+
+async function searchingCallstackForDispatch(
+  pauseFrames: PauseFrame[],
+  replayClient: ReplayClientInterface,
+  sourcesState: SourcesState
+) {
+  // The first 2 elements in filtered pause frames are from replay's redux stub, so they should be ignored
+  // The 3rd element is the user function that calls it, and will most likely be the `dispatch` call
+  let preferredFrameIdx = 2;
+  const applyMiddlwareDecl = await applyMiddlwareDeclCache.readAsync(replayClient, sourcesState);
+
+  for (let frameIdx = 2; frameIdx < pauseFrames.length; frameIdx++) {
+    const frame = pauseFrames[frameIdx];
+
+    if (isFrameInDecl(applyMiddlwareDecl, frame)) {
+      // this is the frame inside `applyMiddleware` where the initial dispatch occurs
+      // the frame just before this one is the user `dispatch`
+      preferredFrameIdx = frameIdx + 1;
+      return preferredFrameIdx;
+    }
+  }
+
+  return null;
+}
+
+async function isReduxMiddleware(sourceContents: string, location: Location) {
+  const sourceOutline = await sourceOutlineCache.readAsync(replayClient, location.sourceId);
+  const dispatchFn = findFunctionOutlineForLocation(location, sourceOutline);
+  const functions = sourceOutline.functions;
+
+  if (dispatchFn) {
+    const index = functions.indexOf(dispatchFn);
+    if (index >= 2) {
+      const wrapDispatch = functions[index - 1];
```

Comment:
:hand: I can think of an edge case where this fails:

```ts
function myMiddleware(storeApi) {
  const someLookupTable = {};

  function someHelper() {}

  return (next) => {
    const anotherVar = 123;
    function anotherFunction() {}

    return (action) => {
      // pause here or whatever
    }
  }
}
```

These are closures with their own scopes, so it's not uncommon to have variables and possibly more functions in scope to do part of the work.

As a more practical example, see this socket handling middleware: https://gist.github.com/markerikson/3df1cf5abbac57820a20059287b4be58 .

I think the defining characteristics we care about are:

- Is it triply nested
- Do the inner two functions have 1 arg ( `next` and `action`) - I can't think of a time when you wouldn't declare both of those, although the names could easily be different
- The outermost function should either have 1 arg or 0 args (ie, it could be declared as `store`, `storeApi`, `{dispatch, getState}`, or `()` with no arg)

Reply:
The latest commit should fix this edge case. Instead of directly considering the function at the position just before the current one as the parent, it now checks for all functions before it until it finds one that is its parent